### PR TITLE
Windows exception handling.

### PR DESCRIPTION
**Pull request by marsupial**
Allows Interpreter to throw and catch exceptions from Interpreter and process.

### CODE REVIEW

**Comment by karies** [comment]
Wow. Any chance of upstreaming most of that?

I did merge this, though - thanks a *lot*!

**Comment by marsupial** [comment]
I'm currently trying to get the fundamental blocks into LLVM.
Then I'll work on an example for them.

**Comment by karies** [comment]
Yippee! Thanks a lot, Roman!

**Comment by bellenot** [comment]
FYI, last week there was something like 4/5 test failures, and Today there are 17 failures on Appveyor, and on my Windows machine there are 30 failures in Release mode and 60 in Debug mode...

**Comment by marsupial** [comment]
I'm aware. Its related to using ELF as all the others are.
If you're not planning on supporting exceptions on Windows then I guess you could revert this, otherwise I would posit these failures previously existed but were just being ignored.

As of yesterday I got cling with COFF to pass all tests.

**Comment by bellenot** [comment]
OK, thanks for the info. We'll discuss what to do next...

**Comment by marsupial** [comment]
Yes let me know, as it stands I want to submit 2 more Windows patches:
A. Deals with Windows static variable functions, but also fixes using C++11 **thread_local** keyword on Mac & Linux.

B. Deals with MSVC runtime registration of std::facets that makes cling crash on exit whenever one is used (like std::end)

Any preference?

**Comment by bellenot** [comment]
BTW, I'm not sure what you mean by "I would posit these failures previously existed but were just being ignored"...

**Comment by bellenot** [comment]
ping... How can I be sure that "these failures previously existed"? Since the tests were passing, would that mean the test infrastructure is broken?

**Comment by bellenot** [comment]
OK, thanks. so I'll try to apply the PR #17 on my machine and I'll let you know how it goes...

**Comment by bellenot** [comment]
BTW, did you try to build in 32 bit?

"C:\ci\build\builddir\tools\cling\tools\driver\cling.vcxproj" (default target) (1) ->
"C:\ci\build\builddir\tools\cling\lib\Utils\clingUtils.vcxproj" (default target) (109) ->
"C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj" (default target) (110) ->
(ClCompile target) ->
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(752): error C2065: 'PRUNTIME_FUNCTION': undeclared identifier [C:\c
i\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(752): error C2146: syntax error: missing ';' before identifier 'RFu
nc' [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(752): error C2065: 'RFunc': undeclared identifier [C:\ci\build\buil
ddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(752): error C2061: syntax error: identifier 'PRUNTIME_FUNCTION' [C:
\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(753): error C2065: 'RUNTIME_FUNCTION': undeclared identifier [C:\ci
\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(760): error C2065: 'PRUNTIME_FUNCTION': undeclared identifier [C:\c
i\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(760): error C2146: syntax error: missing ';' before identifier 'It'
 [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(760): error C2143: syntax error: missing ';' before '=' [C:\ci\buil
d\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(760): error C2143: syntax error: missing ')' before '=' [C:\ci\buil
d\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(760): error C2059: syntax error: '=' [C:\ci\build\builddir\tools\cl
ing\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(760): error C2065: 'It': undeclared identifier [C:\ci\build\builddi
r\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(760): error C2065: 'End': undeclared identifier [C:\ci\build\buildd
ir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(760): error C2059: syntax error: ')' [C:\ci\build\builddir\tools\cl
ing\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(761): error C2065: 'It': undeclared identifier [C:\ci\build\builddi
r\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(761): error C2227: left of '->BeginAddress' must point to class/str
uct/union/generic type [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(761): error C2672: 'std::min': no matching overloaded function foun
d [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(761): error C2780: '_Ty std::min(std::initializer_list<_Elem>,_Pr)'
: expects 2 arguments - 1 provided [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(761): error C2780: 'const _Ty &std::min(const _Ty &,const _Ty &,_Pr
)': expects 3 arguments - 1 provided [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(761): error C2784: '_Ty std::min(std::initializer_list<_Elem>)': co
uld not deduce template argument for 'std::initializer_list<_Elem>' from 'DWORD' [C:\ci\build\builddir\tools\cling\lib\Utils\obj.
clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(761): error C2780: 'const _Ty &std::min(const _Ty &,const _Ty &)':
expects 2 arguments - 1 provided [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(761): error C2059: syntax error: ';' [C:\ci\build\builddir\tools\cl
ing\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(762): error C2065: 'It': undeclared identifier [C:\ci\build\builddi
r\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(762): error C2227: left of '->EndAddress' must point to class/struc
t/union/generic type [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(762): error C2672: 'std::max': no matching overloaded function foun
d [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(762): error C2780: '_Ty std::max(std::initializer_list<_Elem>,_Pr)'
: expects 2 arguments - 1 provided [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(762): error C2780: 'const _Ty &std::max(const _Ty &,const _Ty &,_Pr
)': expects 3 arguments - 1 provided [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(762): error C2784: '_Ty std::max(std::initializer_list<_Elem>)': co
uld not deduce template argument for 'std::initializer_list<_Elem>' from 'DWORD' [C:\ci\build\builddir\tools\cling\lib\Utils\obj.
clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(762): error C2780: 'const _Ty &std::max(const _Ty &,const _Ty &)':
expects 2 arguments - 1 provided [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(764): error C2059: syntax error: 'else' [C:\ci\build\builddir\tools
\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(764): error C2143: syntax error: missing ';' before '{' [C:\ci\buil
d\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(764): error C2447: '{': missing function header (old-style formal l
ist?) [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(772): error C2039: 'RtlAddFunctionTable': is not a member of '`glob
al namespace'' [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(772): error C2065: 'RFunc': undeclared identifier [C:\ci\build\buil
ddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(772): error C2065: 'N': undeclared identifier [C:\ci\build\builddir
\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(772): error C2065: 'BaseAddr': undeclared identifier [C:\ci\build\b
uilddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(772): error C4430: missing type specifier - int assumed. Note: C++
does not support default-int [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(772): error C2440: 'initializing': cannot convert from 'initializer
 list' to 'int' [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(780): error C2039: 'RtlDeleteFunctionTable': is not a member of '`g
lobal namespace'' [C:\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(780): error C2061: syntax error: identifier 'PRUNTIME_FUNCTION' [C:
\ci\build\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(892): error C2059: syntax error: '}' [C:\ci\build\builddir\tools\cl
ing\lib\Utils\obj.clingUtils.vcxproj]
  C:\ci\build\cling-src\tools\cling\lib\Utils\PlatformWin.cpp(892): error C2143: syntax error: missing ';' before '}' [C:\ci\buil
d\builddir\tools\cling\lib\Utils\obj.clingUtils.vcxproj]

    4 Warning(s)
    41 Error(s)

**Comment by marsupial** [comment]
Not in a while.
MS changed how they implemented exception handling on x86 is different than x64.
Getting x64 working was difficult enough, and I have no need to support 32 bit.

File: lib/Utils/PlatformWin.cpp [hunk 1]
```diff
@@ -699,6 +701,192 @@ std::string Demangle(const std::string& Symbol) {
...
+
+  ::RtlAddFunctionTable(RFunc, N, BaseAddr);
+}
+
+void DeRegisterEHFrames(uint8_t* Addr, size_t Size) {
+  assert(getImageBaseMap().find(DWORD64(Addr)) != getImageBaseMap().end());
+
+  ImageBaseMap& Unwind = getImageBaseMap();
+  Unwind.erase(Unwind.find(DWORD64(Addr)));
+  ::RtlDeleteFunctionTable(reinterpret_cast<PRUNTIME_FUNCTION>(Addr));
```
Comment (vgvassilev) [inline lib/Utils/PlatformWin.cpp]: Likewise.

File: lib/Utils/PlatformWin.cpp [hunk 2]
```diff
@@ -699,6 +701,192 @@ std::string Demangle(const std::string& Symbol) {
...
+    }
+  } else {
+    for (PRUNTIME_FUNCTION It = RFunc, End = RFunc +N; It < End; ++It)
+      Ranges.emplace_back(It->BeginAddress, It->EndAddress);
+
+    // Initial sort and merge
+    MergeRanges(Ranges);
+  }
+
+  ::RtlAddFunctionTable(RFunc, N, BaseAddr);
```
Comment (vgvassilev) [inline lib/Utils/PlatformWin.cpp]: @Axel-Naumann, @marsupial, this doesn't compile on win32.

**Comment by marsupial** [comment]
As stated above I am aware of this and there are no plans to implement 32 bit Windows support on this end.

https://msdn.microsoft.com/en-us/library/windows/desktop/ms680659(v=vs.85).aspx

**Comment by vgvassilev** [comment]
It still would make sense to use cling on win 32 bit with exceptions off.

**Comment by marsupial** [comment]
I have two opinions on this:
1. Start small, and get Win64 to actually pass completely before tackling Win32.
2. cling is currently broken in serious ways in regards to exceptions, and compiling it on should at least be the default if not a requirement. Otherwise you [currently] have undefined / non-conformant behavior in a few ways.

**Comment by vgvassilev** [comment]
I am not sure I understand how keeping 32 bit compiling affects 1. For 2. this needs to be seen, the if the user wants to compile cling with exceptions off and run things that's his choice.

I'd suggest fixing win 32 bit compilation for various reasons. One of them is that we may attract somebody like you interested in helping out with 32 bit.

**Comment by marsupial** [comment]
I don't understand:
> For 2. this needs to be seen

The point is I can write a lot of code that does compile but should never be run...unless there is defined behavior your inviting users to use something whose behavior you're unaware of.

Either way, if this is something you feel strongly about it is easily remedied...

**Pull request closed without merging**